src/Microsoft.NetCore.Analyzers/Core/Performance/UsePropertyInsteadOfCountMethodWhenAvailable.cs

```diff
@@ -0,0 +1,226 @@
+                        if (@interface.OriginalDefinition is INamedTypeSymbol originalInterfaceDefinition &&
+                            isCollectionOfTInterface(originalInterfaceDefinition))
+                        {
+                            return true;
+                        }
+                    }
+                }
+                else
+                {
+                    foreach (var @interface in invocationTarget.AllInterfaces)
+                    {
+                        if (@interface.OriginalDefinition is INamedTypeSymbol originalInterfaceDefinition &&
+                            isCollectionOfTInterface(originalInterfaceDefinition))
+                        {
+                            if (invocationTarget.FindImplementationForInterfaceMember(@interface.GetMembers(CountPropertyName)[0]) is IPropertySymbol propertyImplementation &&
+                                !propertyImplementation.ExplicitInterfaceImplementations.Any())
+                            {
+                                return true;
+                            }
+                        }
+                    }
+                }
+
+                return false;
+
+                bool isCollectionOfTInterface(ITypeSymbol type)
+                    => this.ICollectionOfTType.Equals(type.OriginalDefinition);
+            }
+
+            internal bool IsEnumerableType(ISymbol symbol)
+                => this.EnumerableType.Equals(symbol);
+        }
+
+        /// <summary>
+        /// Handler for operaction actions.
+        /// </summary>
+        protected abstract class OperationActionsHandler
+        {
+            protected OperationActionsHandler(OperationActionsContext context)
+            {
+                Context = context;
+            }
+
+            protected OperationActionsContext Context { get; }
+
+            internal void AnalyzeInvocationOperation(OperationAnalysisContext context)
+            {
+                var invocationOperation = (IInvocationOperation)context.Operation;
+
+                if (GetEnumerableCountInvocationTargetType(invocationOperation) is ITypeSymbol invocationTarget &&
+                    GetReplacementProperty(invocationTarget) is string propertyName)
+                {
+                    var propertiesBuilder = ImmutableDictionary.CreateBuilder<string, string>();
+                    propertiesBuilder.Add(PropertyNameKey, propertyName);
+
+                    var diagnostic = Diagnostic.Create(
+                        s_rule,
+                        invocationOperation.Syntax.GetLocation(),
+                        propertiesBuilder.ToImmutable(),
+                        propertyName);
+
+                    context.ReportDiagnostic(diagnostic);
+                }
+            }
+
+            protected abstract ITypeSymbol GetEnumerableCountInvocationTargetType(IInvocationOperation invocationOperation);
+
+            private string GetReplacementProperty(ITypeSymbol invocationTarget)
```

Comment:
Your current approach to determine replacement property is fine. I would even be okay with us being very loose about the requirement here and have all the below checks in this method replaced with a single check that walks the entire type chain of the invocation target and return true if any type has a public int property named either `Length` or `Count`.